### PR TITLE
[c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42944 [c10d][jit] initial torchbind bindings for ProcessGroup
* #47368 [WIP] use lambda to register custom class init
* #47343 [c10d] switch ProcessGroup to be managed by intrusive_ptr
* **#47075 [c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr**
* #47074 [c10d] switch Store to be managed by intrusive_ptr
* #44046 [c10d] switch ProcessGroup::Work to be managed by intrusive_ptr

Differential Revision: [D24667127](https://our.internmc.facebook.com/intern/diff/D24667127)

### CODE REVIEW

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -971,16 +971,17 @@ that adds a prefix to each key inserted to the store.
                   const c10::intrusive_ptr<::c10d::Store>&,
                   int,
                   int,
-                  ::c10d::ProcessGroupNCCL::Options>(),
+                  const c10::intrusive_ptr<
+                      ::c10d::ProcessGroupNCCL::Options>&>(),
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/init.cpp:975]: this is a reference of the ptr, so it means someone else need it keep it alive?

**Review by mrshenli — COMMENTED** [review]
LGTM! Just have one question regarding the lifetime of options.

File: torch/csrc/distributed/c10d/init.cpp (hunk 1 above)
Comment (wanchaol) [reply torch/csrc/distributed/c10d/init.cpp:975]: yes, I think the caller should keep it alive. Specifically I think this option is only needed during the initialization, and we don't need to share the ownership in the `ProcessGroupNCCL` instantiation, as it only record (copy) the fields `opTimeOut` and `isHighPriorityStream`. So after construction, the `Options` is not needed anymore. 

We can also pass by pointer directly, but I didn't see if this is necessary.

**Review by pritamdamania87 — APPROVED** [review]

File: torch/csrc/distributed/c10d/init.cpp (hunk 1 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/c10d/init.cpp:975]: For extra safety, is it fine to remove the reference to the pointer and just incur a refcount bump here?

Comment (wanchaol) [reply torch/csrc/distributed/c10d/init.cpp:975]: yep that's fine as well, I will update the PR to make this a refcount bump to ensure safety.

**Pull request closed without merging**